src/CaptureFile/CaptureFile.cpp

```diff
@@ -262,12 +272,17 @@ ErrorMessageOr<uint64_t> CaptureFileImpl::AddUserDataSection(uint64_t section_si
     }
   }
 
+  // If there are additional sections, the section list needs to be the last section, so it can be
+  // amended.
+  if (!section_list.empty() && IsThereSectionWithOffsetAfterSectionList()) {
+    return ErrorMessage{
+        "Cannot add USER_DATA section - there are sections behind the section list"};
+  }
+
   uint64_t section_list_offset = header_.section_list_offset;
 
-  // If we don't have any additional sections or if there are any sections starting after
-  // section_list move section list to the end of the file.
-  if (header_.section_list_offset == 0 || IsThereSectionWithOffsetAfterSectionList()) {
-    ORBIT_CHECK(section_list.empty());
```

Comment:
This previous CHECK is a little bit wired for me. So previously if we have section after section list, then we assume the section list should be empty?

Reply:
The new logic is much clear than before!